opm/simulators/wells/WellAndGroupStates.hpp

```diff
@@ -0,0 +1,564 @@
+    }
+    const std::vector<WellState>& wellStates() const
+    {
+        return well_states_;
+    }
+
+    std::map<std::string, GroupState>& groupStates()
+    {
+        return group_states_;
+    }
+    const std::map<std::string, GroupState>& groupStates() const
+    {
+        return group_states_;
+    }
+
+
+private:
+
+    // -----------  Data members  -----------
+
+    std::vector<WellState> well_states_;
+    std::vector<std::string> well_names_;
+    std::map<std::string, GroupState> group_states_;
+
+    std::vector<std::vector<PerforationData>> well_perf_data_;
+
+
+    // -----------  Private functions  -----------
+
+    template <unsigned long N>
+    static double sum(const std::array<double, N>& x)
+    {
+        return std::accumulate(x.begin(), x.end(), 0.0);
+    }
+
+    static void initSingleWell(const std::vector<double>& cell_pressures,
+                               const Schedule& schedule,
+                               const Well& well,
+                               const int report_step,
+                               const PhaseUsage& phase_usage,
+                               const std::vector<PerforationData>& perf_data,
+                               const SummaryState& summary_state,
+                               SingleWellState<NumActivePhases>& wstate)
+    {
+        const auto inj_controls = well.isInjector() ? well.injectionControls(summary_state) : Well::InjectionControls(0);
+        const auto prod_controls = well.isProducer() ? well.productionControls(summary_state) : Well::ProductionControls(0);
+
+        // Initialize flag, statuses and current controls.
+        initFlagsAndControls(schedule, well, report_step, inj_controls, prod_controls, wstate);
+
+        // Initialize bhp, thp and well rates.
+        initPressuresAndRates(cell_pressures, well, phase_usage, perf_data, inj_controls, prod_controls, wstate);
+
+        // Initialize connection pressures and rates.
+        initConnections(cell_pressures, perf_data, wstate);
+
+        // Initialize multi-segment well parts (the 'segments' member).
+        if (well.isMultiSegment()) {
+            initMultiSegment(well, wstate);
+        }
+    }
+
+
+
+
+    static void initSingleWellFromPrevious(const SingleWellState<NumActivePhases>& prev_wstate,
+                                           SingleWellState<NumActivePhases>& wstate)
+    {
+        // Flags and statuses.
+        wstate.status = prev_wstate.status;
+        wstate.is_producer = prev_wstate.is_producer;
+        wstate.effective_events_occurred = prev_wstate.effective_events_occurred;
+        wstate.current_injection_control = prev_wstate.current_injection_control;
+        wstate.current_production_control = prev_wstate.current_production_control;
+
+        // Quantities.
+        wstate.bhp = prev_wstate.bhp;
+        wstate.thp = prev_wstate.thp;
+        wstate.temperature = prev_wstate.temperature; // 20 degrees Celcius by default.
```

Comment:
Comment not interesting in this context.

Reply:
Agree...